File: src/main/java/com/xin/util/ClassByteUtil.java
package com.xin.util;

import com.xin.base.GeneralClassAdapter;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;

import static org.objectweb.asm.ClassWriter.COMPUTE_MAXS;

/**
 * @author linxixin@cvte.com
 * @since 1.0
 */
public class ClassByteUtil {

    public static byte[] getNewClassBytes(String className, ClassLoader classLoader, GeneralClassAdapter method) {

        ClassReader classReader = new ClassReader(getClassByte(classLoader, className));
        ClassWriter classWriter = new ClassWriter(classReader, COMPUTE_MAXS);
        method.setCp(classWriter);
        classReader.accept(method, ClassReader.EXPAND_FRAMES);
        return classWriter.toByteArray();
    }

    public static byte[] getClassByte(ClassLoader classLoader, String name) {
        try (InputStream resourceAsStream = classLoader.getResourceAsStream(name.replace('.', '/') + ".class")) {

            if (resourceAsStream == null) {
                throw new RuntimeException(name + " 类不存在无法进行注入");
            }
            return readClass(resourceAsStream);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static byte[] readClass(InputStream resourceAsStream) {
        final ByteArrayOutputStream output = new ByteArrayOutputStream();
        try (BufferedInputStream bufferedInputStream = new BufferedInputStream(resourceAsStream)) {
            int n;
            byte[] buffer = new byte[1024 * 2];
            while (true) {
                if (-1 == (n = bufferedInputStream.read(buffer))) {
                    break;
                }
                output.write(buffer, 0, n);
            }
            return output.toByteArray();
        } catch (IOException e) {
            throw new RuntimeException("没读取到文件");
        }
    }

}

File: src/main/java/com/xin/util/ClassByteUtil.kt
package com.xin.util

import com.xin.base.GeneralClassAdapter
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.ClassWriter.COMPUTE_MAXS
import vo.BaseException
import java.io.IOException
import java.io.InputStream

/**
 *
 * @author linxixin@cvte.com
 */

fun getNewClassBytes(className: String, classLoader: ClassLoader, method: GeneralClassAdapter): ByteArray {
    val classReader = ClassReader(getClassByte(classLoader, className))
    val classWriter = ClassWriter(classReader, COMPUTE_MAXS or COMPUTE_MAXS)
    method.setCp(classWriter)
    classReader.accept(method, ClassReader.EXPAND_FRAMES)
    return classWriter.toByteArray()
}

fun getClassByte(classLoader: ClassLoader, name: String): ByteArray {
    try {
        val resourceAsStream = classLoader.getResourceAsStream(name.replace('.', '/') + ".class")
                ?: throw BaseException(name + " 类不存在无法进行注入")
        return readClass(
                resourceAsStream, true)
    } catch (e: IOException) {
        e.printStackTrace()
    }

    return ByteArray(0)
}

fun readClass(inputStream: InputStream, close: Boolean): ByteArray {
    try {
        var b = ByteArray(inputStream.available())
        var len = 0
        while (true) {
            val n = inputStream.read(b, len, b.size - len)
            if (n == -1) {
                if (len < b.size) {
                    val c = ByteArray(len)
                    System.arraycopy(b, 0, c, 0, len)
                    b = c
                }
                return b
            }
            len += n
            if (len == b.size) {
                val last = inputStream.read()
                if (last < 0) {
                    return b
                }
                val c = ByteArray(b.size + 1000)
                System.arraycopy(b, 0, c, 0, len)
                c[len++] = last.toByte()
                b = c
            }
        }
    } finally {
        if (close) {
            inputStream.close()
        }
    }
}
